edb/edgeql/compiler/schemactx.py

```diff
@@ -552,40 +548,34 @@ def apply_intersection(
         # of the argument, then this is, effectively, a NOP.
         return TypeIntersectionResult(stype=left)
 
-    is_subtype = False
-    empty_intersection = False
-    union = left.get_union_of(ctx.env.schema)
-    if union:
-        # If the argument type is a union type, then we
-        # narrow it by the intersection type.
-        narrowed_union = []
-        for component_type in union.objects(ctx.env.schema):
-            if component_type.issubclass(ctx.env.schema, right):
-                narrowed_union.append(component_type)
-            elif right.issubclass(ctx.env.schema, component_type):
-                narrowed_union.append(right)
-
-        if len(narrowed_union) == 0:
-            int_type = get_intersection_type((left, right), ctx=ctx)
-            is_subtype = int_type.issubclass(ctx.env.schema, left)
-            assert isinstance(right, s_obj.InheritingObject)
-            empty_intersection = not any(
-                c.issubclass(ctx.env.schema, left)
-                for c in right.descendants(ctx.env.schema)
-            )
-        elif len(narrowed_union) == 1:
-            int_type = narrowed_union[0]
-            is_subtype = int_type.issubclass(ctx.env.schema, left)
-        else:
-            int_type = get_union_type(narrowed_union, ctx=ctx)
-    else:
-        is_subtype = right.issubclass(ctx.env.schema, left)
-        empty_intersection = not is_subtype
-        int_type = get_intersection_type((left, right), ctx=ctx)
+    if right.issubclass(
+        ctx.env.schema, left, check_opaque_unions=True, check_views=True
+    ):
+        # The intersection type is a proper *subclass* and can be directly
+        # narrowed.
+        return TypeIntersectionResult(
+            stype=right,
+            is_empty=False,
+            is_subtype=True,
+        )
+
+    if (
+        left.get_is_opaque_union(ctx.env.schema)
+        and (left_union := left.get_union_of(ctx.env.schema))
+    ):
+        left = get_union_type(left_union.objects(ctx.env.schema), ctx=ctx)
```

Comment:
The opaque union stuff really is dodgy in general. Does this represent a change of behavior, or are we doing basically what we were before for opaque unions?

Reply:
In particular, I think that intersecting with an opaque union really *ought* to just produce the intersector type---but the reason we don't want that is for implementation purposes, IIRC?

Reply:
Always producing intersections here causes issues when compiling the standard library because it tries to create an intersection `TypeShell`.